apps/emqx_bridge_pulsar/src/emqx_bridge_pulsar_impl_producer.erl

```diff
@@ -0,0 +1,399 @@
+%% `emqx_resource' API
+%%-------------------------------------------------------------------------------------
+
+callback_mode() -> async_if_possible.
+
+%% there are no queries to be made to this bridge, so we say that
+%% buffer is supported so we don't spawn unused resource buffer
+%% workers.
+is_buffer_supported() -> true.
+
+-spec on_start(manager_id(), config()) -> {ok, state()}.
+on_start(InstanceId, Config) ->
+    #{
+        authentication := _Auth,
+        bridge_name := BridgeName,
+        servers := Servers0,
+        ssl := SSL
+    } = Config,
+    Servers = format_servers(Servers0),
+    ClientId = make_client_id(InstanceId, BridgeName),
+    SSLOpts = emqx_tls_lib:to_client_opts(SSL),
+    ClientOpts = #{
+        ssl_opts => SSLOpts,
+        conn_opts => conn_opts(Config)
+    },
+    case pulsar:ensure_supervised_client(ClientId, Servers, ClientOpts) of
+        {ok, _Pid} ->
+            ?SLOG(info, #{
+                msg => "pulsar_client_started",
+                instance_id => InstanceId,
+                pulsar_hosts => Servers
+            });
+        {error, Reason} ->
+            ?SLOG(error, #{
+                msg => "failed_to_start_pulsar_client",
+                instance_id => InstanceId,
+                pulsar_hosts => Servers,
+                reason => Reason
+            }),
+            throw(failed_to_start_pulsar_client)
+    end,
+    start_producer(Config, InstanceId, ClientId, ClientOpts).
+
+-spec on_stop(manager_id(), state()) -> ok.
+on_stop(_InstanceId, State) ->
+    #{
+        pulsar_client_id := ClientId,
+        producers := Producers
+    } = State,
+    stop_producers(ClientId, Producers),
+    stop_client(ClientId),
+    ?tp(pulsar_bridge_stopped, #{instance_id => _InstanceId}),
+    ok.
+
+-spec on_get_status(manager_id(), state()) -> connected | disconnected.
+on_get_status(_InstanceId, State) ->
+    #{
+        pulsar_client_id := ClientId,
+        producers := Producers
+    } = State,
+    case pulsar_client_sup:find_client(ClientId) of
+        {ok, Pid} ->
+            try pulsar_client:get_status(Pid) of
+                true ->
+                    get_producer_status(Producers);
+                false ->
+                    disconnected
+            catch
+                error:timeout ->
+                    disconnected;
+                exit:{noproc, _} ->
+                    disconnected
+            end;
+        {error, _} ->
+            disconnected
+    end.
+
+-spec on_query(manager_id(), {send_message, map()}, state()) -> ok | {error, timeout}.
+on_query(_InstanceId, {send_message, Message}, State) ->
+    #{
+        producers := Producers,
+        sync_timeout := SyncTimeout,
+        message_template := MessageTemplate
+    } = State,
+    PulsarMessage = render_message(Message, MessageTemplate),
+    try
+        pulsar:send_sync(Producers, [PulsarMessage], SyncTimeout)
```

Comment:
The pulsar lib says that `send_sync` return type is `{ok, _Response} | {error, term()}`, but `ok` is specified

Reply:
Good catch, I've updated the typespec.

:heavy_check_mark: